From b565ea80a2a1e20ae542dec85f8e8494afe561fa Mon Sep 17 00:00:00 2001
From: Ruslan Bay <ruslanbey@proton.me>
Date: Thu, 26 Feb 2026 17:21:29 +0100
Subject: [PATCH 39/60] media: intel/ipu4: Guard cleanup against uninitialized
 resources

Add NULL pointer checks during driver cleanup to prevent dereferencing
partially-initialized resources when initialization fails mid-way.
Guard against uninitialized video queues and V4L2 subdevices.
---
 drivers/media/pci/intel/ipu-isys-csi2-be-soc.c | 3 +++
 drivers/media/pci/intel/ipu-isys-csi2-be.c     | 3 +++
 drivers/media/pci/intel/ipu-isys-queue.c       | 3 +++
 drivers/media/pci/intel/ipu-isys-tpg.c         | 3 +++
 drivers/media/pci/intel/ipu-isys-video.c       | 3 +++
 drivers/media/pci/intel/ipu4/ipu4-isys-isa.c   | 3 +++
 6 files changed, 18 insertions(+)

diff --git a/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c b/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
index ab2570f25..4d788b653 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
@@ -236,6 +236,9 @@ void ipu_isys_csi2_be_soc_cleanup(struct ipu_isys_csi2_be_soc *csi2_be_soc)
 {
 	int i;
 
+	if (!csi2_be_soc->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&csi2_be_soc->asd.sd);
 	ipu_isys_subdev_cleanup(&csi2_be_soc->asd);
 	for (i = 0; i < NR_OF_CSI2_BE_SOC_STREAMS; i++)
diff --git a/drivers/media/pci/intel/ipu-isys-csi2-be.c b/drivers/media/pci/intel/ipu-isys-csi2-be.c
index bbf1b3ff5..2c1fc7123 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2-be.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2-be.c
@@ -206,6 +206,9 @@ static void csi2_be_set_ffmt(struct v4l2_subdev *sd,
 
 void ipu_isys_csi2_be_cleanup(struct ipu_isys_csi2_be *csi2_be)
 {
+	if (!csi2_be->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&csi2_be->asd.sd);
 	ipu_isys_subdev_cleanup(&csi2_be->asd);
 	ipu_isys_video_cleanup(&csi2_be->av);
diff --git a/drivers/media/pci/intel/ipu-isys-queue.c b/drivers/media/pci/intel/ipu-isys-queue.c
index d95bee52f..14c9dc84e 100644
--- a/drivers/media/pci/intel/ipu-isys-queue.c
+++ b/drivers/media/pci/intel/ipu-isys-queue.c
@@ -1415,5 +1415,8 @@ int ipu_isys_queue_init(struct ipu_isys_queue *aq)
 
 void ipu_isys_queue_cleanup(struct ipu_isys_queue *aq)
 {
+	if (!aq->vbq.ops)
+		return;
+
 	vb2_queue_release(&aq->vbq);
 }
diff --git a/drivers/media/pci/intel/ipu-isys-tpg.c b/drivers/media/pci/intel/ipu-isys-tpg.c
index 59151f395..10aeec19d 100644
--- a/drivers/media/pci/intel/ipu-isys-tpg.c
+++ b/drivers/media/pci/intel/ipu-isys-tpg.c
@@ -251,6 +251,9 @@ static struct media_entity_operations tpg_entity_ops = {
 
 void ipu_isys_tpg_cleanup(struct ipu_isys_tpg *tpg)
 {
+	if (!tpg->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&tpg->asd.sd);
 	ipu_isys_subdev_cleanup(&tpg->asd);
 	ipu_isys_video_cleanup(&tpg->av);
diff --git a/drivers/media/pci/intel/ipu-isys-video.c b/drivers/media/pci/intel/ipu-isys-video.c
index 7c1a740f8..2e3476803 100644
--- a/drivers/media/pci/intel/ipu-isys-video.c
+++ b/drivers/media/pci/intel/ipu-isys-video.c
@@ -2294,6 +2294,9 @@ int ipu_isys_video_init(struct ipu_isys_video *av,
 
 void ipu_isys_video_cleanup(struct ipu_isys_video *av)
 {
+	if (!av->aq.vbq.ops)
+		return;
+
 	video_unregister_device(&av->vdev);
 	media_entity_cleanup(&av->vdev.entity);
 	mutex_destroy(&av->mutex);
diff --git a/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c b/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
index 4ac6545ec..a9652b66c 100644
--- a/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
+++ b/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
@@ -231,6 +231,9 @@ static struct media_entity_operations isa_entity_ops = {
 
 void ipu_isys_isa_cleanup(struct ipu_isys_isa *isa)
 {
+	if (!isa->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&isa->asd.sd);
 	ipu_isys_subdev_cleanup(&isa->asd);
 	ipu_isys_video_cleanup(&isa->av_scaled);
-- 
2.55.0

